kernel/kernel-api/src/main/java/io/delta/kernel/internal/checksum/ChecksumUtils.java

```diff
@@ -169,6 +201,149 @@ private static CRCInfo buildCrcInfoWithFullLogReplay(
         Optional.of(state.addedFileSizeHistogram));
   }
 
+  /**
+   * Attempts to build CRC info incrementally from the last seen checksum. Falls back if incremental
+   * computation is not possible.
+   *
+   * @param lastSeenCrcInfo The last available CRC info to build upon
+   * @param engine The engine to use for file operations
+   * @param logSegment The log segment to process
+   * @return Optional containing the new CRC info, or empty if fallback is needed
+   */
+  private static Optional<CRCInfo> buildCrcInfoIncrementally(
+      CRCInfo lastSeenCrcInfo, Engine engine, LogSegment logSegment) throws IOException {
+    // Can only build incrementally if we have domain metadata and file size histogram
+    if (!lastSeenCrcInfo.getDomainMetadata().isPresent()) {
+      logger.info("Falling back to full replay: detected current crc missing domain metadata.");
+      return Optional.empty();
+    }
+    if (!lastSeenCrcInfo.getFileSizeHistogram().isPresent()) {
+      logger.info("Falling back to full replay: detected current crc missing file size histogram.");
+      return Optional.empty();
+    }
+
+    // Initialize state tracking
+    StateTracker state = new StateTracker();
+
+    // Create iterator for delta files newer than last CRC
+    try (CloseableIterator<ColumnarBatch> iterator =
+        DeltaLogActionUtils.readCommitFiles(
+            engine,
+            logSegment.getDeltas().stream()
```

Comment:
can we use the minor compacted logs? If yes that can be a separate PR

Reply:
Good point, added TODO, I need to look into details of different cases of minor compacted logs